workspaces/arborist/lib/query-selector-all.js

```diff
@@ -0,0 +1,512 @@
+        [...node.edgesIn].some(edge => edge.dev)
+      )
+    },
+    async '.optional' (prevResults) {
+      return prevResults.filter(node =>
+        [...node.edgesIn].some(edge => edge.optional)
+      )
+    },
+    async '.peer' (prevResults) {
+      return prevResults.filter(node =>
+        [...node.edgesIn].some(edge => edge.peer)
+      )
+    },
+    async '.workspace' (prevResults) {
+      return prevResults.filter(node =>
+        node.isWorkspace
+      )
+    },
+    async '.bundled' (prevResults) {
+      return prevResults.filter(node =>
+        node.inBundle
+      )
+    },
+  }))
+
+  const hasParent = async (node, compareNodes) => {
+    if (parentCache.has(node) && parentCache.get(node).has(compareNodes)) {
+      return true
+    }
+    const parentFound = compareNodes.some(compareNode => {
+      // follows logical parent for link anscestors
+      if (node.isTop && (node.resolveParent === compareNode)) {
+        return true
+      }
+      // follows edges-in to check if they match a possible parent
+      return [...node.edgesIn].some(edge => edge && edge.from === compareNode)
+    })
+
+    if (parentFound) {
+      if (!parentCache.has(node)) {
+        parentCache.set(node, new Set())
+      }
+      parentCache.get(node).add(compareNodes)
+    }
+
+    return parentFound
+  }
+
+  // checks if a given node is a descendant of any
+  // of the nodes provided in the compare nodes array
+  const hasAscendant = async (node, compareNodes) => {
+    const hasP = await hasParent(node, compareNodes)
+    if (hasP) {
+      return true
+    }
+
+    const lookupEdgesIn = async (node) => {
+      const edgesIn = [...node.edgesIn]
+      const p = await Promise.all(
+        edgesIn.map(edge =>
+          edge && edge.from && hasAscendant(edge.from, compareNodes))
+      )
+      return edgesIn.some((edge, index) => p[index])
+    }
+    const ancestorFound = (node.isTop && node.resolveParent)
+      ? await hasAscendant(node.resolveParent, compareNodes)
+      : await lookupEdgesIn(node)
+
+    return ancestorFound
+  }
+
+  const combinatorsMap = new Map(Object.entries({
+    async '>' (prevResults, nextResults) {
+      const p = await Promise.all(
+        nextResults.map(i => hasParent(i, prevResults))
+      )
+      return nextResults.filter((nextItem, index) => p[index])
+    },
+    async ' ' (prevResults, nextResults) {
+      const p = await Promise.all(
+        nextResults.map(i => hasAscendant(i, prevResults))
+      )
+      return nextResults.filter((nextItem, index) => p[index])
+    },
+    async '~' (prevResults, nextResults) {
+      const p = await Promise.all(nextResults.map(nextItem => {
+        const seenNodes = new Set()
+        const possibleParentNodes =
+          prevResults
+            .flatMap(node => {
+              seenNodes.add(node)
+              return [...node.edgesIn]
+            })
+            .map(edge => edge.from)
+            .filter(Boolean)
```

Comment:
```suggestion
            .flatMap(edge => [edge.from] || [])
```

Reply:
That won't work cause we'd just wrap the falsey edge.from in an array which is always truthy
```
> !![null]
true
```

Reply:
I think we just want `.flatMap(edge => edge.from || [])`